### PR TITLE
Check the sizes of the inputs are compatible with the shapes of the placeholders they fill

### PR DESCRIPTION
**Pull request by oxinabox**
No-one would put the wrong size into a placeholder intensionly.
And we already have mechanisms to allow parts of a placeholder to have flexible sizes.

Putting the wrong sized input can render uses of shape inference incorrect.
So we should not let people do it.

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -151,10 +151,16 @@ function TensorShape(dims::Vector)
     TensorShape([x<0 ? Nullable{Int64}() : Nullable{Int64}(x) for x in dims])
 end
 
-function TensorShape(dim::Void)
+function TensorShape(::typeof(collect(tuple())))
```
Comment (malmaud) [inline src/core.jl]: What does this complicated type mean? Is there is a more compact type signature that would work?

File: examples/logistic.jl [hunk 2]
```diff
@@ -19,20 +20,22 @@ y = draw(y_prob)
 
 # Build the model
 sess = Session(Graph())
-X = placeholder(Float64, shape=[nothing, 10])
-Y_obs = placeholder(Float64, shape=[100])
+@tf begin
```
Comment (malmaud) [inline examples/logistic.jl]: Can we make this change to the example a separate PR?

File: src/core.jl (hunk 1 above)
Comment (oxinabox) [inline src/core.jl]: Yes, could instead say: `Vector{Union{}}`.

I was playing with the idea, that `(::typeof(collect(tuple())` might be clearer.
Since it does not require remembering that collecting some untyped (not Any-typed or Void typed) collection gives `Vector{Union{}}`.

I will make the change

File: examples/logistic.jl (hunk 2 above)
Comment (oxinabox) [inline examples/logistic.jl]: The example does need to change, as it uses inconstant sizes; and is a test, which would otherwise fail.

But I can remove the `@tf` because I only added that to help debug, which line had the inconstant sizing.

Comment (oxinabox) [inline examples/logistic.jl]: done

File: src/core.jl (hunk 1 above)
Comment (oxinabox) [inline src/core.jl]: done
